File: app/src/main/scala/io/github/yueeng/hacg/WebActivity.scala
package io.github.yueeng.hacg

import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v4.widget.SwipeRefreshLayout.OnRefreshListener
import android.support.v7.app.AppCompatActivity
import android.view._
import android.webkit.{WebChromeClient, WebView, WebViewClient}
import android.widget.ProgressBar
import io.github.yueeng.hacg.Common._
import io.github.yueeng.hacg.ViewBinder.ViewBinder

class WebActivity extends AppCompatActivity {
  override def onCreate(savedInstanceState: Bundle): Unit = {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_web)

    setSupportActionBar(findViewById(R.id.toolbar))
    getSupportActionBar.setLogo(R.mipmap.ic_launcher)
    getSupportActionBar.setDisplayHomeAsUpEnabled(true)

    val transaction = getSupportFragmentManager.beginTransaction()

    val fragment = getSupportFragmentManager.findFragmentById(R.id.container) match {
      case fragment: WebFragment => fragment
      case _ => new WebFragment().arguments(getIntent.getExtras)
    }

    transaction.replace(R.id.container, fragment)

    transaction.commit()
  }

  override def onBackPressed(): Unit = {
    getSupportFragmentManager.findFragmentById(R.id.container) match {
      case f: WebFragment if f.web.canGoBack => f.web.goBack()
      case _ => super.onBackPressed()
    }
  }

  override def onOptionsItemSelected(item: MenuItem): Boolean = {
    item.getItemId match {
      case android.R.id.home => super.onBackPressed(); true
      case _ => super.onOptionsItemSelected(item)
    }
  }
}

class WebFragment extends Fragment {
  val busy = new ViewBinder[Boolean, SwipeRefreshLayout](false)((view, value) => view.post(() => view.setRefreshing(value)))
  var uri: String = _

  def defuri: String = getArguments match {
    case b: Bundle if b.containsKey("url") => b.getString("url")
    case _ => HAcg.philosophy
  }

  lazy val web: WebView = getView.findViewById(R.id.web)
  lazy val progress: ProgressBar = getView.findViewById(R.id.progress)

  override def onCreateOptionsMenu(menu: Menu, inflater: MenuInflater): Unit = {
    inflater.inflate(R.menu.menu_web, menu)
    super.onCreateOptionsMenu(menu, inflater)
  }

  override def onOptionsItemSelected(item: MenuItem): Boolean = {
    item.getItemId match {
      case R.id.open => Common.openWeb(getActivity, uri); true
      case _ => super.onOptionsItemSelected(item)
    }
  }

  override def onCreate(state: Bundle): Unit = {
    super.onCreate(state)
    setHasOptionsMenu(true)
    setRetainInstance(true)
    uri = if (state != null && state.containsKey("url")) state.getString("url") else defuri
  }

  override def onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle): View = {
    val root = inflater.inflate(R.layout.fragment_web, container, false)
    val web: WebView = root.findViewById(R.id.web)

    busy += root.findViewById(R.id.swipe)
    busy.views.head.setOnRefreshListener(new OnRefreshListener {
      override def onRefresh(): Unit = web.loadUrl(uri)
    })

    val settings = web.getSettings
    settings.setJavaScriptEnabled(true)
    settings.setUseWideViewPort(true)
    settings.setLoadWithOverviewMode(true)
    val back = root.findViewById[View](R.id.button2)
    val fore = root.findViewById[View](R.id.button3)
    web.setWebViewClient(new WebViewClient() {
      override def shouldOverrideUrlLoading(view: WebView, url: String): Boolean = {
        view.loadUrl(url)
        true
      }

      override def onPageStarted(view: WebView, url: String, favicon: Bitmap): Unit = {
        busy <= true
        progress.setProgress(0)
      }

      override def onPageFinished(view: WebView, url: String): Unit = {
        busy <= false
        progress.setProgress(100)
        uri = url

        back.setEnabled(view.canGoBack)
        fore.setEnabled(view.canGoForward)
      }
    })
    web.setWebChromeClient(new WebChromeClient() {
      override def onProgressChanged(view: WebView, newProgress: Int): Unit = {
        progress.setProgress(newProgress)
      }
    })

    val click = viewClick {
      _.getId match {
        case R.id.button1 => web.loadUrl(defuri)
        case R.id.button2 => if (web.canGoBack) web.goBack()
        case R.id.button3 => if (web.canGoForward) web.goForward()
        case R.id.button4 => web.loadUrl(uri)
      }
    }
    List(R.id.button1, R.id.button2, R.id.button3, R.id.button4)
      .map(root.findViewById[View]).foreach(_.setOnClickListener(click))

    web.loadUrl(uri)
    root
  }

  override def onSaveInstanceState(state: Bundle): Unit = {
    state.putString("url", uri)
  }
}

File: app/src/main/java/io/github/yueeng/hacg/WebActivity.kt
package io.github.yueeng.hacg

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.os.Bundle
import android.view.*
import android.webkit.CookieManager
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.appcompat.app.AppCompatActivity
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.AbstractSavedStateViewModelFactory
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.savedstate.SavedStateRegistryOwner
import io.github.yueeng.hacg.databinding.ActivityWebBinding
import io.github.yueeng.hacg.databinding.FragmentWebBinding

class WebActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(ActivityWebBinding.inflate(layoutInflater).also { binding ->
            setSupportActionBar(binding.toolbar)
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }.root)

        supportFragmentManager.beginTransaction()
                .replace(R.id.container, supportFragmentManager.findFragmentById(R.id.container)
                        ?.let { it as? WebFragment }
                        ?: WebFragment().arguments(intent.extras))
                .commit()
    }

    override fun onBackPressed() {
        supportFragmentManager.findFragmentById(R.id.container)
                ?.let { it as? WebFragment }
                ?.takeIf { it.onBackPressed() }
                ?: super.onBackPressed()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> {
            super.onBackPressed()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }
}

class WebViewModel(handle: SavedStateHandle, args: Bundle?) : ViewModel() {
    val busy = handle.getLiveData("busy", false)
    val uri = handle.getLiveData("url", args?.getString("url")!!)
}

class WebViewModelFactory(owner: SavedStateRegistryOwner, private val defaultArgs: Bundle? = null) : AbstractSavedStateViewModelFactory(owner, defaultArgs) {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel?> create(key: String, modelClass: Class<T>, handle: SavedStateHandle): T = WebViewModel(handle, defaultArgs) as T
}

class WebFragment : Fragment() {
    private val viewModel: WebViewModel by viewModels { WebViewModelFactory(this, bundleOf("url" to defuri)) }

    private val defuri: String
        get() = arguments?.takeIf { it.containsKey("url") }?.getString("url")
                ?: (if (isLogin) "${HAcg.philosophy}?foro=signin" else HAcg.philosophy)
    private val isLogin: Boolean get() = arguments?.getBoolean("login", false) ?: false

    fun onBackPressed(): Boolean {
        val binding = FragmentWebBinding.bind(requireView())
        if (binding.web.canGoBack()) {
            binding.web.goBack()
            return true
        }
        return false
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_web, menu)
        return super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.open -> {
            activity?.openUri(viewModel.uri.value!!)
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        setHasOptionsMenu(true)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        FragmentWebBinding.bind(requireView()).web.destroy()
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val binding = FragmentWebBinding.inflate(inflater, container, false)
        CookieManager.getInstance().acceptThirdPartyCookies(binding.web)
        viewModel.busy.observe(viewLifecycleOwner, { binding.swipe.isRefreshing = it })
        binding.swipe.setOnRefreshListener { binding.web.loadUrl(viewModel.uri.value!!) }

        val settings = binding.web.settings
        settings.javaScriptEnabled = true
        settings.useWideViewPort = true
        settings.loadWithOverviewMode = true
        val back = binding.button2
        val fore = binding.button3
        binding.web.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView?, url: String): Boolean =
                    activity?.openUri(url, false) == true

            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                viewModel.busy.postValue(true)
                binding.progress.progress = 0
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                viewModel.busy.postValue(false)
                binding.progress.progress = 100
                viewModel.uri.postValue(url)

                back.isEnabled = view?.canGoBack() ?: false
                fore.isEnabled = view?.canGoForward() ?: false
                if (isLogin) view?.evaluateJavascript("""favorites_data["user_id"]""") { s ->
                    s?.trim('"')?.toIntOrNull()?.takeIf { it != 0 }?.let {
                        user = it
                        activity!!.finish()
                    }
                }
            }
        }
        binding.web.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, newProgress: Int) {
                binding.progress.progress = newProgress
            }
        }

        val click = View.OnClickListener { v ->
            when (v?.id) {
                R.id.button1 -> binding.web.loadUrl(defuri)
                R.id.button2 -> if (binding.web.canGoBack()) binding.web.goBack()
                R.id.button3 -> if (binding.web.canGoForward()) binding.web.goForward()
                R.id.button4 -> binding.web.loadUrl(viewModel.uri.value!!)
            }
        }
        listOf(binding.button1, binding.button2, binding.button3, binding.button4)
                .forEach { it.setOnClickListener(click) }

        binding.web.loadUrl(viewModel.uri.value!!)
        return binding.root
    }
}
